src/__/collections/filter.php

```diff
@@ -28,14 +49,31 @@
  * [['name' => 'fred', 'age' => 32]]
  * ```
  *
- * @param array         $array   Array to filter
- * @param \Closure|null $closure Closure to filter the array
+ * @since 0.2.0 iterable objects are now supported
  *
- * @return array
+ * @param iterable      $iterable Array to filter
+ * @param \Closure|null $closure  Closure to filter the array
+ *
+ * @throws \InvalidArgumentException when an non-array or non-traversable object is given for $iterable.
+ *
+ * @return array|\Generator When given a `\Traversable` object for `$iterable`, a generator will be returned.
+ *                          Otherwise, an array will be returned.
  */
-function filter(array $array, \Closure $closure = null)
+function filter($iterable, \Closure $closure = null)
 {
-    return \array_values(
-        $closure ? \array_filter($array, $closure) : \array_filter($array)
-    );
+    if (is_array($iterable)) {
+        $values = [];
```

Comment:
Could be factorized with `iterator_to_array(filterIterable($iterable, $closure))`?

Reply:
Yup! Good call, I missed this one :sweat_smile:

Reply:
Done in 94d6cf8ee04e3fd9c5b849ab1b70d3a6a976a98b